### PR TITLE
Replace confd & rsyslog with custom code

### PR DESCRIPTION
**Pull request by jakolehm**
This PR replaces confd and rsyslog with custom ruby implementation.

- certificate split/validation: `CertSplitter`
- etcd polling: `Actors::EtcdWatcher`
- haproxy config generation: `Actors::HaproxyConfigGenerator`
- haproxy process management: `Actors::HaproxySpawner`
- haproxy syslog handling: `Actors::SyslogServer`

Haproxy templates are splitted into multiple files (for readability) and rendered using excellent `Hanami::View`.

Todo:

- [x] Remove confd
- [x] Remove rsyslog
- [x] Fix failing tests (not too many)
- [x] Rspec

### CODE REVIEW

**Comment by jakolehm** [comment]
E2E (bats) test results:

```
$ bats test/
 ✗ basic auth gives 401 without user and password
   (in test file test/basic_auth_test.bats, line 17)
     `[ "${lines[0]}" = "401" ]' failed
     200
   www.foo.com
   service-b:9292
   user admin insecure-password passwd
 ✓ basic auth gives 200 with valid user and password
 ✓ basic auth gives 200 with valid user and password, password encrypted
 ✓ ciphers are set
 ✓ supports cookie stickyness
 ✓ supports cookie stickyness with custom cookie config
 ✓ supports cookie stickyness with custom cookie prefix
 ✓ supports custom common settings via env
 ✓ returns custom error page
 ✓ returns health check page if configured in env
 ✗ returns error if health not configured in env
   (in test file test/health_checks_test.bats, line 22)
     `[ $(expr "$output" : ".*503 — Service Unavailable.*") -ne 0 ]' failed
   Error:  100: Key not found (/kontena/haproxy/lb/services/service-a) [827]
   Error:  100: Key not found (/kontena/haproxy/lb/services/service-b) [827]
   Error:  100: Key not found (/kontena/haproxy/lb/services/service-c) [827]
 ✓ supports health check uri setting for balanced service
 ✓ redirects *.foo.com -> foo.com
 ✓ supports ssl with invalid cert ignored
 ✓ supports virtual_hosts
 ✓ supports wildcard virtual_hosts
 ✓ supports virtual_hosts + virtual_path
 ✓ supports virtual_hosts + virtual_path + keep_virtual_path
 ✓ handles empty upstreams
 ✓ on duplicate virtual_hosts first one in alphabets wins
 ✗ prioritizes first vhost+vpath, then vhost and finally vpath
   (in test file test/virtual_host_test.bats, line 151)
     `[ "${lines[0]}" = "service-b" ]' failed
     service-a
   Error:  100: Key not found (/kontena/haproxy/lb/services/service-a) [875]
   /
   service-a:9292
   www.bar.com
   service-b:9292
   api.bar.com
   /c
   service-c:9292
 ✓ works with domain:port host header
 ✓ supports virtual_path
 ✓ supports virtual_path + keep_virtual_path

24 tests, 3 failures
```

File: Dockerfile [hunk 1]
```diff
@@ -1,28 +1,25 @@
...
-    apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 505D97A41C61B9CD && \
-    apt-get update
-
-RUN apt-get install -yq --no-install-recommends haproxy ca-certificates curl net-tools rsyslog && \
-    apt-get clean && \
-    rm -rf /var/lib/apt/lists/* && \
-    curl -sL -o /bin/confd https://github.com/kelseyhightower/confd/releases/download/v${CONFD_VERSION}/confd-${CONFD_VERSION}-linux-amd64 && \
-    chmod +x /bin/confd && \
-    curl -sL -o /bin/tini https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini && \
+RUN apk update && apk --update add curl bash tzdata ruby ruby-irb ruby-bigdecimal \
```
Comment (jnummelin) [inline Dockerfile]: Why `curl`, for possible debugging purposes?

File: lib/kontena/actors/etcd_watcher.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  class EtcdWatcher < Concurrent::Actor::RestartingContext
+    include Kontena::Logging
+
+    attr_reader :client, :path
+
+    def initialize(host, path)
+      info "initialized with etcd host=#{host} path=#{path}"
+      @host = host
+      @path = path
+      @read_task = Concurrent::TimerTask.new(execution_interval: 1) {
```
Comment (jnummelin) [inline lib/kontena/actors/etcd_watcher.rb:14]: Couldn't we use also etcd watch with correct path for this?

File: lib/kontena/actors/haproxy_spawner.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    def validate_config
+      system(@validate_cmd.join(' ')) == true
+    end
+
+    def reload_haproxy
+      info 'Reloading HAProxy'
+      reload_cmd = @haproxy_cmd + ['-sf', @current_pid.to_s]
+      pid = Process.spawn(reload_cmd.join(' '))
+      Process.wait(@current_pid)
```
Comment (jnummelin) [inline lib/kontena/actors/haproxy_spawner.rb]: Why do we wait for the previous haproxy process to end? Since hproxy is "reloaded" with `-sf` flag, it could take really long for the process to actually exit as it waits till all the connections are terminated.

File: lib/kontena/templates/haproxy/http_backends.text.erb [hunk 4]
```diff
@@ -0,0 +1,42 @@
+<% services.each do |service| %>
+backend <%= service.name %>
+    option forwardfor
+    <% service.custom_settings.each do |setting| %>
+    <%= setting %>
+    <% end %>
+
+    <% if service.cookie %>
+    <% if service.cookie.empty? %>
```
Comment (jnummelin) [inline lib/kontena/templates/haproxy/http_backends.text.erb:9]: Styling: could there be some spacing to more easily identify the nesting ifs?

**Review by jnummelin — COMMENTED** [review]
Looks really good already. Not sure if I understand the concurrent-ruby actor model deeply enough to comment on those specifics though.

There seems to be few bats cases still failing:
```
 ✗ returns custom error page
   (in test file test/error_page_test.bats, line 15)
     `[ $(expr "$output" : ".*Kontena Load Balancer.*") -ne 0 ]' failed
   Error:  100: Key not found (/kontena/haproxy/lb/services/service-a) [1864]
   Error:  100: Key not found (/kontena/haproxy/lb/services/service-c) [1865]
 ✗ returns health check page if configured in env
   (in test file test/health_checks_test.bats, line 16)
     `[ $(expr "$output" : ".*Everything seems to be 200 - OK.*") -ne 0 ]' failed
   Error:  100: Key not found (/kontena/haproxy/lb/services/service-a) [1865]
   Error:  100: Key not found (/kontena/haproxy/lb/services/service-b) [1865]
   Error:  100: Key not found (/kontena/haproxy/lb/services/service-c) [1865]
 ✗ returns error if health not configured in env
   (in test file test/health_checks_test.bats, line 22)
     `[ $(expr "$output" : ".*503 — Service Unavailable.*") -ne 0 ]' failed
   Error:  100: Key not found (/kontena/haproxy/lb/services/service-a) [1865]
   Error:  100: Key not found (/kontena/haproxy/lb/services/service-b) [1865]
   Error:  100: Key not found (/kontena/haproxy/lb/services/service-c) [1865]
```
Also few rspecs failing:
```
rspec ./spec/kontena/views/http_in_spec.rb:5 # Kontena::Views::HttpIn.render balance sets balance to roundrobin by default
rspec ./spec/kontena/views/http_in_spec.rb:16 # Kontena::Views::HttpIn.render balance sets balance to configured value
rspec ./spec/kontena/views/http_in_spec.rb:43 # Kontena::Views::HttpIn.render cookies adds default cookie policy if cookie value is empty string
rspec ./spec/kontena/views/http_in_spec.rb:57 # Kontena::Views::HttpIn.render cookies adds custom cookie policy if cookie value is not empty string
rspec ./spec/kontena/views/http_in_spec.rb:73 # Kontena::Views::HttpIn.render basic auth adds basic auth config
```

File: lib/kontena/actors/etcd_watcher.rb (hunk 2 above)
Comment (jakolehm) [reply lib/kontena/actors/etcd_watcher.rb:14]: Could, but it can trigger massive amount of changes.

File: Dockerfile (hunk 1 above)
Comment (jakolehm) [reply Dockerfile]: Dockerfile fetches tini using curl.

Comment (jnummelin) [reply Dockerfile]: Do'h, me blind 🙈

**Comment by jakolehm** [comment]
@jnummelin @SpComb PTAL

File: lib/kontena/actors/etcd_watcher.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    def start
+      @read_task.execute
+    end
+
+    def client
+      @client ||= Etcd::Client.new(host: @host, port: 2379)
+    end
+
+    def read_etcd
+      response = client.get(path, recursive: true)
```
Comment (jnummelin) [inline lib/kontena/actors/etcd_watcher.rb:43]: Do we always need to read the full recursive content? I think there could be a check on the etcd `index` counter as to indicate if anything has really changed.

File: lib/kontena/models/common_service.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
+module Kontena::Models
+  module CommonService
+
+    attr_accessor :name,
+                  :upstreams,
+                  :balance,
+                  :health_check_uri,
```
Comment (jnummelin) [inline lib/kontena/models/common_service.rb]: Is `health_check_uri` really common?

File: lib/kontena/templates/haproxy/index.text.erb [hunk 7]
```diff
@@ -0,0 +1,39 @@
+global
```
Comment (jnummelin) [inline lib/kontena/templates/haproxy/index.text.erb]: All the other "partials" are named based on the HAProxy config "context", shoud this file also follow that convention and be named as `global.text.erb`? `index` is IMO bit confusing name :)

**Review by jnummelin — COMMENTED** [review]
Overall looks good, added few minor comments.

Needs more "real-life" testing, maybe we should build as push `kontena/lb:next` or something so community could help testing this also.

File: lib/kontena/actors/etcd_watcher.rb (hunk 5 above)
Comment (jakolehm) [reply lib/kontena/actors/etcd_watcher.rb:43]: Probably no need to read full content, but let's optimize this later.

File: Dockerfile [hunk 8]
```diff
@@ -1,28 +1,25 @@
-FROM ubuntu-debootstrap:trusty
+FROM haproxy:1.5-alpine
```
Comment (jnummelin) [inline Dockerfile:1]: Should we upgrade haproxy also? Or maybe leave that as separate PR :)

File: Dockerfile [hunk 9]
```diff
@@ -1,28 +1,25 @@
...
 MAINTAINER Kontena, Inc. <info@kontena.io>
 
-ENV CONFD_VERSION=0.11.0 \
-    STATS_PASSWORD=secret \
-    TINI_VERSION=v0.8.4
+ENV STATS_PASSWORD=secret \
+    TINI_VERSION=v0.14.0 \
```
Comment (jnummelin) [inline Dockerfile]: `0.16.1` seems to be latest. Changelog doesn't show anything major to be happened, but maybe still good to go with latest.

File: lib/kontena/actors/etcd_watcher.rb [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+      Concurrent.global_io_executor
+    end
+
+    # @param [Message] msg
+    def on_message(msg)
+      case msg.action
+      when :start
+        start
+      else
+        pass
```
Comment (jnummelin) [inline lib/kontena/actors/etcd_watcher.rb]: This should fail on the senders context (raise?) when trying to send something else than `:start`?

File: lib/kontena/actors/lb_supervisor.rb [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        pass
+      end
+    end
+
+    def on_message_symbol(action)
+      case action
+      when :reset
+        info "got reset event"
+      else
+        info "unhandled message symbol: #{action}"
```
Comment (jnummelin) [inline lib/kontena/actors/lb_supervisor.rb]: Why not `pass`?

File: lib/kontena/actors/haproxy_config_generator.rb [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+      services
+    end
+
+    # @param [Etcd::Node] node
+    # @return [Kontena::Models::Service]
+    def generate_service(node)
+      root = node
+      service = Kontena::Models::Service.new(node.key.split('/')[-1])
+      node.children.each do |c|
+        if c.key == "#{root.key}/upstreams"
```
Comment (jnummelin) [inline lib/kontena/actors/haproxy_config_generator.rb]: For every key we do string ops :/  maybe just grab the last part of the key and compare that against some freezed strings?

**Comment by SpComb** [comment]
How does this relate to https://github.com/kontena/kontena-haproxy? Could the two be merged?

File: lib/kontena/actors/haproxy_config_generator.rb (hunk 12 above)
Comment (jnummelin) [reply lib/kontena/actors/haproxy_config_generator.rb]: Not sure which reads better, `if/elsif` or `case/when` :)

Comment (jnummelin) [reply lib/kontena/actors/haproxy_config_generator.rb]: Also thinking if this initialization logic would be better on the `Kontena::Models::Service` model

File: lib/kontena/actors/etcd_watcher.rb (hunk 2 above)
Comment (SpComb) [reply lib/kontena/actors/etcd_watcher.rb:14]: The [`Kontena::Etcd::Reader`](https://github.com/SpComb/kontena-etcd/blob/master/lib/kontena/etcd/reader.rb) is intended to provide that functionality.

File: lib/kontena/actors/haproxy_spawner.rb [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+          reload_haproxy
+        else
+          start_haproxy
+        end
+      end
+    end
+
+    def start_haproxy
+      info 'Starting HAProxy process'
+      @current_pid = Process.spawn(@haproxy_cmd.join(' '))
```
Comment (SpComb) [inline lib/kontena/actors/haproxy_spawner.rb]: Something must also be a `Process.wait` once haproxy exits? That happens after every reload, and this will leak haproxy zombies?

There should be a `SIGCHLD` handler somewhere?

File: lib/kontena/actors/lb_supervisor.rb [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+    def update_haproxy
+      @spawner << Message.new(:update)
+    end
+
+    def on_event(event)
+      case event.class
+      when Concurrent::Actor::UnknownMessage
+        info "on unknown message event: #{event.reference.inspect}"
+      else
+        info "on event: #{event}"
```
Comment (SpComb) [inline lib/kontena/actors/lb_supervisor.rb]: Just debugging? Is this supposed to be overriden?

File: lib/kontena/actors/haproxy_spawner.rb [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+    end
+
+    def validate_config
+      system(@validate_cmd.join(' ')) == true
```
Comment (SpComb) [inline lib/kontena/actors/haproxy_spawner.rb]: Would prefer the shell-less `system(haproxy_bin, '-f', config_file, '-t')` (also for `Process.spawn`) instead of the shell-joining-and-splitting, but it doesn't make much of a difference here because there's nothing that needs quoting.

File: lib/kontena/actors/lb_supervisor.rb [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+        update_haproxy
+      else
+        pass
+      end
+    end
+
+    def on_message_symbol(action)
+      case action
+      when :reset
+        info "got reset event"
```
Comment (SpComb) [inline lib/kontena/actors/lb_supervisor.rb]: Probably not used?

File: lib/kontena/actors/lb_supervisor.rb (hunk 11 above)
Comment (SpComb) [inline lib/kontena/actors/lb_supervisor.rb]: Should this `pass` instead?

File: lib/kontena/actors/haproxy_config_generator.rb [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+
+    # @param [Etcd::Node] node
+    # @return [Array<Kontena::Models::Service]
+    def generate_services(node)
+      services = []
+      node.children.sort_by { |c| c.key }.each do |c|
+        services << generate_service(c)
+      end
+
+      services
```
Comment (SpComb) [inline lib/kontena/actors/haproxy_config_generator.rb:49]: Also `freeze` the array?

File: lib/kontena/actors/haproxy_config_generator.rb [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+    def generate_tcp_services(node)
+      services = []
+      node.children.sort_by { |c| c.key }.each do |c|
+        service = generate_tcp_service(c)
+        if service.upstreams.size > 0 && service.external_port
+          services << service
+        end
+      end
+
+      services
```
Comment (SpComb) [inline lib/kontena/actors/haproxy_config_generator.rb]: Also `freeze` the array?

File: lib/kontena/actors/haproxy_config_generator.rb [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+          services = generate_services(c)
+        elsif c.key == "#{root}/tcp-services"
+          tcp_services = generate_tcp_services(c)
+        end
+      end
+
+      config = Kontena::Views::Haproxy.render({
+        format: :text, services: services, tcp_services: tcp_services
+      }).each_line.reject{ |l| l.strip == ''.freeze }.join
+      parent << Message.new(:write_config, config)
```
Comment (SpComb) [inline lib/kontena/actors/haproxy_config_generator.rb]: Hmm... if the models are just rendered out to a string, do they need to get frozen?

File: lib/kontena/actors/haproxy_spawner.rb [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+    end
+
+    def update_haproxy
+      if validate_config
+        if current_pid
+          reload_haproxy
+        else
+          start_haproxy
+        end
+      end
```
Comment (SpComb) [inline lib/kontena/actors/haproxy_spawner.rb:38]: Should there be some `else` here, to report errors on invalid configs? Or those get passed through from the `system` call `stderr`?

**Review by SpComb — COMMENTED** [review]
The general actor architecture looks reasonable, although I suspect I might want some more logging to see what's going on when it runs.

:+1: for the immutable/frozen models, although the `HaproxyConfigGenerator` actor seems a little off... it doesn't actually do anything concurrently, and it passes out a rendered string, so the nice immutable/frozen models don't offer concurrent access :(

Although... looking at the `Kontena::Actors::LbSupervisor`, it's maybe not the best example of an actor-based architecture... it currently does everything in serial... it's mostly equivalent to a single-threaded loop that runs the same pipeline...? Might get some actual concurrency once you throw in the `HaproxySpawner` `SIGCHLD` handling.

Need to do a dynamic analysis next (actually run the code)

File: lib/kontena/actors/etcd_watcher.rb [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+      @read_task.execute
+    end
+
+    def client
+      @client ||= Etcd::Client.new(host: @host, port: 2379)
+    end
+
+    def read_etcd
+      response = client.get(path, recursive: true)
+      self.parent << Message.new(:generate_config, response)
```
Comment (jnummelin) [inline lib/kontena/actors/etcd_watcher.rb]: This makes things async, right? This starts a chain of events: `generate_config` -> `write_config` -> `update_haproxy`. If some of the steps take a while could the order of updates be messed up? I mean later update going on the fast-lane and update first to haproxy.

File: lib/kontena/templates/haproxy/http_backends.text.erb [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+
+    <% if service.basic_auth? %>
+    acl auth_ok_<%= service.name %> http_auth(auth_users_<%= service.name %>)
+    http-request auth realm <%= service.name %> unless auth_ok_<%= service.name %>
+    <% end %>
+
+    <% if service.virtual_path? %>
+    <% if !service.keep_virtual_path? %>
+    reqrep ^([^\ :]*)\ <%= service.virtual_path %>[/]?(.*)     \1\ /\2
+    <% end %>
```
Comment (jnummelin) [inline lib/kontena/templates/haproxy/http_backends.text.erb]: Somehow makes me feel there should be some `else` branch also? If not, then just combine those if's?

File: test/error_page_test.bats [hunk 23]
```diff
@@ -10,6 +10,7 @@ setup() {
 }
 
 @test "returns custom error page" {
+  etcdctl set /kontena/haproxy/lb/services/service-b/virtual_hosts www.foo.com
```
Comment (jnummelin) [inline test/error_page_test.bats:13]: Why is this needed?

File: lib/kontena/actors/etcd_watcher.rb (hunk 21 above)
Comment (jakolehm) [reply lib/kontena/actors/etcd_watcher.rb]: Yes, it's async but I don't see how it could mess up order of updates?

File: lib/kontena/actors/etcd_watcher.rb (hunk 10 above)
Comment (jakolehm) [reply lib/kontena/actors/etcd_watcher.rb]: Yeah, I can change it to raise error.

File: lib/kontena/actors/haproxy_config_generator.rb (hunk 19 above)
Comment (jakolehm) [reply lib/kontena/actors/haproxy_config_generator.rb]: Not sure what you mean?

File: lib/kontena/actors/etcd_watcher.rb (hunk 10 above)
Comment (jakolehm) [reply lib/kontena/actors/etcd_watcher.rb]: Actually no. It will `pass` message to next behaviour which is `Behaviour::ErrorsOnUnknownMessage`.

File: lib/kontena/actors/etcd_watcher.rb (hunk 21 above)
Comment (SpComb) [reply lib/kontena/actors/etcd_watcher.rb]: Need to understand how the actors behave... do they process messages strictly in serial? If they can't start processing a new message until the previous message has been processed, then this isn't a problem.

File: lib/kontena/actors/haproxy_config_generator.rb (hunk 19 above)
Comment (SpComb) [reply lib/kontena/actors/haproxy_config_generator.rb]: I initially assumed that the point of the model `freeze`s in the `generate_service` etc functions was to prepare the model objects for inclusion in messages sent between actors, but if the models get rendered to a string in the same actor, then that's moot.

I guess they're just to keep the view templates from mutating the models?

File: lib/kontena/actors/haproxy_config_generator.rb (hunk 12 above)
Comment (jakolehm) [reply lib/kontena/actors/haproxy_config_generator.rb]: Optimized for less string operations and switched to `case`.

File: lib/kontena/actors/haproxy_config_generator.rb (hunk 18 above)
Comment (jakolehm) [reply lib/kontena/actors/haproxy_config_generator.rb]: Done

File: lib/kontena/actors/lb_supervisor.rb (hunk 11 above)
Comment (jakolehm) [reply lib/kontena/actors/lb_supervisor.rb]: Maybe, I just wasn't comfortable with crashing top-level supervisor if it gets unknown message.

File: lib/kontena/actors/haproxy_spawner.rb (hunk 13 above)
Comment (jakolehm) [reply lib/kontena/actors/haproxy_spawner.rb]: If haproxy process is reloaded with `-sf` it is allowed to process ongoing connections. So yes, it will leak processes until they drop all connections (and this is by design). Container also has `tini` running as process reaper.

We could probably improve this logic so that those old processes would be killed after a while (no point to keep them forever).

File: lib/kontena/actors/haproxy_spawner.rb (hunk 20 above)
Comment (jakolehm) [reply lib/kontena/actors/haproxy_spawner.rb:38]: Haproxy logs validation errors and we catch those via syslog.

File: lib/kontena/actors/lb_supervisor.rb (hunk 14 above)
Comment (jakolehm) [reply lib/kontena/actors/lb_supervisor.rb]: Just debugging and it's ok to override this. I can remove it.

File: lib/kontena/templates/haproxy/http_backends.text.erb (hunk 4 above)
Comment (jakolehm) [reply lib/kontena/templates/haproxy/http_backends.text.erb:9]: I optimized this for output. Not sure how to make both output & code clean :/

File: lib/kontena/templates/haproxy/http_backends.text.erb (hunk 22 above)
Comment (jakolehm) [reply lib/kontena/templates/haproxy/http_backends.text.erb]: This is direct port from current (confd) template. I can combine those if's.

File: test/error_page_test.bats (hunk 23 above)
Comment (jakolehm) [reply test/error_page_test.bats:13]: Because it does not listen http port anymore without configured http-services.

File: lib/kontena/actors/haproxy_config_generator.rb (hunk 19 above)
Comment (jakolehm) [reply lib/kontena/actors/haproxy_config_generator.rb]: Yeah, I'm not 💯 sure that rendering them in the same actor makes actually sense. Probably rendering should be done somewhere else (would make pluggable datasource easy). IMHO this refactoring can be done later.

File: lib/kontena/actors/etcd_watcher.rb (hunk 10 above)
Comment (jnummelin) [reply lib/kontena/actors/etcd_watcher.rb]: I guess the question was more that what does `Behaviour::ErrorsOnUnknownMessage` actually do? Will it cause the actor to crash or raise something on the caller side?

File: lib/kontena/actors/etcd_watcher.rb (hunk 21 above)
Comment (jnummelin) [reply lib/kontena/actors/etcd_watcher.rb]: My thinking was something like:
`t` means time, "seconds".
t=0, etcd reader reads and emits the data
t=0.5, `generate_config` invocated (invocation a), for some reason it slows down
t=1, etcd reader reads and emits the data
t=1.5, `generate_config` invocated (invocation b)
t=1.7, `generate_config` invocated (invocation b) finishes and triggers `write_config`
t=1.8, `generate_config` invocated (invocation a) finishes and triggers `write_config`
...
vóila, ordering messed up :)

File: test/error_page_test.bats (hunk 23 above)
Comment (jnummelin) [reply test/error_page_test.bats:13]: hmm, that also means that if I'd configure `KONTENA_LB_HEALTH_URI` to create LB level "healthcheck" it woulnd't work either?

Comment (jakolehm) [reply test/error_page_test.bats:13]: Yes.

Comment (jakolehm) [reply test/error_page_test.bats:13]: If user has only tcp-services, health checks should be done at tcp (port) level. If user has http-services then we should not open http port until we have proper config from etcd.

**Comment by jakolehm** [comment]
@SpComb @jnummelin PTAL (again)

File: test/error_page_test.bats (hunk 23 above)
Comment (jnummelin) [reply test/error_page_test.bats:13]: Fair enough, just wanted to make sure this is a known design choice. :)

File: lib/kontena/templates/haproxy/http_in.text.erb [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+    errorfile 502 /etc/haproxy/errors/502.http
+    errorfile 503 /etc/haproxy/errors/503.http
+
+    <% if ENV['KONTENA_LB_HEALTH_URI'] %>
+    monitor-uri <%= ENV['KONTENA_LB_HEALTH_URI'] %>
+    errorfile 200 /etc/haproxy/errors/200.http
+    <% end %>
+    
+<% services.each do |service| %>
+<% if service.upstreams? %>
```
Comment (SpComb) [inline lib/kontena/templates/haproxy/http_in.text.erb]: This is different vs the equivalent conditional in the current confd templates, because those are actually buggy: #26

**Review by SpComb — COMMENTED** [review]
Looks like this also fixes a ["bug"](https://github.com/kontena/kontena/issues/1725) in the template's upstreams len check, which AFAIK means that the current `kontena/lb` will actually still configure the service frontend ACLs even if the service doesn't have any upstreams.

Per https://github.com/kontena/kontena-loadbalancer/pull/26#issuecomment-301720778 I suspect that might actually be a feature.... if all the backends for a service are down, it will get a HTTP 503 from the lb instead of falling through to any default `virtual_path=/` backend? AFAIK this happens for anything like `kontena service stop`, or temporarily during re-deploys of a single-instance service?

File: lib/kontena/templates/haproxy/http_in.text.erb (hunk 24 above)
Comment (SpComb) [reply lib/kontena/templates/haproxy/http_in.text.erb]: Verified that this also "fixes" the tests in #26, and I'm not sure that's a good thing... modifying the tests to expect the service-b definition to still be there, it's now gone. The current `kontena/lb:latest` code will keep the `use_backend service-b` even if it doesn't have any upstreams:

#### `test/no_upstreams_test.bats`
```sh
#!/usr/bin/env bats

load "common"


setup() {
  etcdctl rm --recursive /kontena/haproxy/lb/services/service-b || true
}


@test "if no upstreams, service frontend retained in config" {
  etcdctl set /kontena/haproxy/lb/services/service-b/virtual_hosts www.foo.com

  sleep 1
  run curl -sL -w "%{http_code}" -H "Host: www.foo.com" http://localhost:8180/ -o /dev/null
  [ "${lines[0]}" = "503" ]

  run config
  assert_output_contains "use_backend service-b" 1
}
```

#### `bats test/no_upstreams_test.bats` (master)
```
 ✓ if no upstreams, service frontend retained in config
```

#### `bats test/no_upstreams_test.bats` (feature/remove-confd-and-rsyslog)
```
 ✗ if no upstreams, service frontend retained in config
   (from function `assert_equal' in file test/common.bash, line 29,
    from function `assert_output_contains' in file test/common.bash, line 39,
    in test file test/no_upstreams_test.bats, line 19)
     `assert_output_contains "use_backend service-b" 1' failed
     global
   www.foo.com
   expected: 1
   actual:   0
```

**Comment by SpComb** [comment]
Also passes the relevant `SSL_CERTS`-related tests in #28 that test with the actual `curl --cacert` without `-k` :+1:

File: lib/kontena/templates/haproxy/http_in.text.erb (hunk 24 above)
Comment (SpComb) [reply lib/kontena/templates/haproxy/http_in.text.erb]: See #29 for a test for this "bug"

File: entrypoint.sh [hunk 25]
```diff
@@ -1,75 +1,14 @@
...
-function wait_confd() {
-  # Loop until confd has updated the haproxy config
-  until confd -onetime -node "$ETCD_NODE" -prefix="/kontena/haproxy/$LB_NAME" "$@"; do
-    echo "[kontena-lb] waiting for confd to refresh haproxy.cfg"
-    sleep 5
-  done
-}
-
-if [ -z "$ETCD_NODE"]; then
+if [ -z "$ETCD_NODE" ]; then
```
Comment (SpComb) [inline entrypoint.sh:9]: Also fixes https://github.com/kontena/kontena/issues/2987

File: lib/kontena/templates/haproxy/http_in.text.erb (hunk 24 above)
Comment (jakolehm) [reply lib/kontena/templates/haproxy/http_in.text.erb]: Maybe we need to revert changes related to that bug fix.

File: lib/kontena/templates/haproxy/http_in.text.erb [hunk 26]
```diff
@@ -0,0 +1,41 @@
+listen http-in
+    bind *:80
+    http-request replace-header Host (.*):.* \1
+    <% if ENV['SSL_CERTS'] %>
+    bind *:443 ssl crt /etc/haproxy/certs/ no-sslv3
```
Comment (SpComb) [inline lib/kontena/templates/haproxy/http_in.text.erb:5]: :+1: Better to use the instance-local `ENV['SSL_CERTS']` for this, rather than going via the etcd `.../certs/bundle` key like the current confd template does.

**Review by SpComb — COMMENTED** [review]
Also fixes https://github.com/kontena/kontena/issues/3019

File: lib/kontena/templates/haproxy/http_in.text.erb (hunk 24 above)
Comment (SpComb) [reply lib/kontena/templates/haproxy/http_in.text.erb]: AFAIK just removing the `<% if service.upstreams? %>` conditional should be enough to make this pass #29

File: lib/kontena/templates/haproxy/tcp_proxies.text.erb [hunk 27]
```diff
@@ -0,0 +1,16 @@
+<% services.each do |service| %>
+<% if service.upstreams? && service.external_port %>
```
Comment (SpComb) [inline lib/kontena/templates/haproxy/tcp_proxies.text.erb]: Remove the `service.upstreams?` conditional to make this pass #29 for TCP services as well

File: lib/kontena/templates/haproxy/http_backends.text.erb [hunk 28]
```diff
@@ -0,0 +1,42 @@
+<% services.each do |service| %>
+<% if service.upstreams? %>
```
Comment (SpComb) [inline lib/kontena/templates/haproxy/http_backends.text.erb]: Remove the `service.upstreams?` conditional to make this pass #29

File: lib/kontena/templates/haproxy/http_in.text.erb [hunk 29]
```diff
@@ -0,0 +1,41 @@
...
+<% end %>
+<% end %>
+
+<% services.select { |s| s.upstreams? && s.virtual_hosts.size > 0 && s.virtual_path }.each do |service| %>
+    use_backend <%= service.name %> if host_<%= service.name %> host_<%= service.name %>_virtual_path
+<% end %>
+<% services.select { |s| s.upstreams? && s.virtual_hosts.size > 0 && s.virtual_path.nil? }.each do |service| %>
+    use_backend <%= service.name %> if host_<%= service.name %>
+<% end %>
+<% services.select { |s| s.upstreams? && s.virtual_path && s.virtual_hosts.size == 0 }.each do |service| %>
```
Comment (SpComb) [inline lib/kontena/templates/haproxy/http_in.text.erb]: Remove all three `s.upstreams?` conditionals to make this pass #29

**Review by SpComb — CHANGES_REQUESTED** [review]
I think this is ready to :white_check_mark: and merge, except for the known feature-not-bug-regression re #29, which should only need a couple trivial changes to remove the `service.upstreams?` checks.

Entirely different question as to when we want to release this as `kontena/lb:latest` and automatically upgrade all existing installations, but AFAIK we're doing the automated image builds from the `v1` branch now, so we can safely merge this to master and then figure out how to handle the release?

**Comment by jakolehm** [comment]
Removed `service.upstreams?` checks.

**Comment by jakolehm** [comment]
```
$ bats test/
 ✓ basic auth gives 401 without user and password
 ✓ basic auth gives 200 with valid user and password
 ✓ basic auth gives 200 with valid user and password, password encrypted
 ✓ ciphers are set
 ✓ supports cookie stickyness
 ✓ supports cookie stickyness with custom cookie config
 ✓ supports cookie stickyness with custom cookie prefix
 ✓ supports custom common settings via env
 ✓ if no upstreams, service frontend retained in config
 ✓ returns custom error page
 ✓ returns health check page if configured in env
 ✓ returns error if health not configured in env
 ✓ supports health check uri setting for balanced service
 ✓ redirects *.foo.com -> foo.com
 ✓ supports ssl with invalid cert ignored
 ✓ supports virtual_hosts
 ✓ supports wildcard virtual_hosts
 ✓ supports virtual_hosts + virtual_path
 ✓ supports virtual_hosts + virtual_path + keep_virtual_path
 ✓ handles empty upstreams
 ✓ on duplicate virtual_hosts first one in alphabets wins
 ✓ prioritizes first vhost+vpath, then vhost and finally vpath
 ✓ works with domain:port host header
 ✓ supports virtual_path
 ✓ supports virtual_path + keep_virtual_path

25 tests, 0 failures
```

**Review by SpComb — APPROVED** [review]
LGTM for merge to `master` to continue development with things like #28, but still need to decide when this is appropriate to release as `kontena/lb:latest`
